packages/server/lib/browsers/protocol.js

```diff
@@ -59,30 +94,13 @@ const getWsTargetFor = (port) => {
     debug('failed to connect to CDP %o', { connectOpts, err })
   })
   .then(() => {
-    debug('CRI.List on port %d', port)
-
-    // what happens if the next call throws an error?
-    // it seems to leave the browser instance open
-    return CRI.List(connectOpts)
-  })
-  .then((targets) => {
-    debug('CRI List %o', { numTargets: targets.length, targets })
-    // activate the first available id
-    // find the first target page that's a real tab
-    // and not the dev tools or background page.
-    // since we open a blank page first, it has a special url
-    const newTabTargetFields = {
-      type: 'page',
-      url: 'about:blank',
-    }
-
-    const target = _.find(targets, newTabTargetFields)
-
-    la(target, 'could not find CRI target')
-
-    debug('found CRI target %o', target)
+    const findPage = findStartPageTarget.bind(null, connectOpts)
 
-    return target.webSocketDebuggerUrl
+    // p-retry https://github.com/sindresorhus/p-retry uses options
+    // from https://github.com/tim-kos/node-retry#retryoperationoptions
+    // and defaults (1 second first retry, exponential factor 2, 10 attempts)
+    // are reasonable
+    return pRetry(findPage)
```

Comment:
the way this is implemented, you won't get the retry counter in stdout once the TCP connection succeeds. this retry loop will also follow completely different rules from the initial retry loop. this behavior is defined and tested for, but this `pRetry` just glosses over it.

i think the easiest way to keep the retry counter and the existing retry behavior would be to just use the existing `createRetryingSocket` and call the `retry` function from the callback to trigger retries as necessary. it will not make the most beautiful code, but it will make the most beautiful user experience :)

Reply:
I doubt this matters much, and if this solves the problem then all good. To me the initial connection seems to take longer than `CRI.List` update

Reply:
It's not the *most* important thing, but still, in this PR we have a chance to...

* maintain a consistent user experience,
* have well-defined retry behavior around an error-prone section of code,
* avoid introducing new dependencies since we already have this functionality in createRetryingSocket,
* and reduce confusion/technical debt around this code section instead of adding new complexity and external libraries that developers have to understand

If we don't do it before this PR gets merged, it just becomes even more tech debt, and this code is only a few months old so we should try to keep it in nice shape for as long as possible.

I can implement it if you like since I already know the APIs involved, just let me know.

Reply:
Be my guest and change it.
I would prefer using off the shelf 3rd party libs for this, since they are well tested. I also don't find the argument for introducing technical debt convincing. We will do more stuff in the retry function:
- connect to socket
- `CRI.List`
- check if the list contains the blank page

If the check fails, we are going to go ... and connect back to the socket? What if we just need more time ... to get `CRI.List` to query the tabs AFTER the connection? Will we get into a loop where we are not solving the problem? 

Right now it seems to me the retries are split where they need to be and are isolated.

Reply:
You raise a good point re: reconnecting to the socket on each retry, that probably shouldn't happen.

I still think that it's important to preserve the retry messaging. Otherwise the error experience will look like this:

```
Retrying 20/33...
Retrying 21/33...
Failed to connect to browser
```

...then the user is left head-scratching, "why didn't it just retry like it said it was going to?" and we'll have to fix it later on...

So maybe we can keep pRetry and then move the "warning message" into a function that keeps track of how many times it's retried, so we can use it in both places when a retry occurs.